File: Java/15. 3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < nums.length - 2; i++) {
            if (i != 0 && nums[i] == nums[i - 1])
                continue;
            twoSum(nums, i + 1, -nums[i], res);    
        }
        return res;
    }
    
    private void twoSum(int[] nums, int start, int target, List<List<Integer>> res) {
        int left = start, right = nums.length - 1;        
        while (left < right) {            
            if (nums[left] + nums[right] == target) {                
                res.add(new ArrayList<>(Arrays.asList(-target, nums[left], nums[right])));
                left += 1;
                right -= 1;
                while (left < right && nums[left] == nums[left - 1])
                    left += 1;
                while (left < right && nums[right] == nums[right + 1])
                    right -= 1;
            } 
            else if (nums[left] + nums[right] > target)
                right -= 1;
            else
                left += 1;
        }        
    }
}

File: Python/15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        ans = list()
        for i in range(0, len(nums) - 2):
            if i == 0 or (i > 0 and nums[i] != nums[i - 1]):
                lo = i + 1
                hi = len(nums) - 1
                while lo < hi:
                    if nums[lo] + nums[hi] == -nums[i]:
                        ans.append([nums[i], nums[lo], nums[hi]])
                        while lo < hi and nums[lo] == nums[lo + 1]:
                            lo += 1
                        while lo < hi and nums[hi] == nums[hi - 1]:
                            hi -= 1
                        lo += 1
                        hi -= 1
                    elif nums[lo] + nums[hi] > -nums[i]:
                        hi -= 1
                    else:
                        lo += 1
        return ans
